sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -121,72 +119,88 @@ void printArgs(Arg0 arg0, Args... args) {
 
 // Utility function to check return from pi calls.
 // Throws if pi_result is not a PI_SUCCESS.
-// TODO: Absorb this utility in Trace Class
-template <typename Exception> inline void piCheckThrow(PiResult pi_result) {
+template <typename Exception = cl::sycl::runtime_error>
+inline void piCheckResult(PiResult pi_result) {
   CHECK_OCL_CODE_THROW(pi_result, Exception);
 }
 
-// Utility function to check if return from pi call is
-// PI_SUCCESS. If is it not, throw a cl::sycl::runtime_error.
-// TODO: Absorb this utility in Trace Class
-inline void piCheckResult(PiResult pi_result) {
-  piCheckThrow<cl::sycl::runtime_error>(pi_result);
-}
+#define PI_TRACE_ENABLED (std::getenv("SYCL_PI_TRACE") != nullptr)
 
-template <typename FnType> class Trace {
+template <typename FnType, size_t FnOffset> class Trace {
 private:
   FnType m_FnPtr;
-  static bool m_TraceEnabled;
+  std::string m_FnName;
 
 public:
-  Trace(FnType FnPtr, const std::string &FnName) : m_FnPtr(FnPtr) {
-    if (m_TraceEnabled)
-      std::cout << "---> " << FnName << "(";
-  }
-
+  Trace();
   template <typename... Args> PiResult operator()(Args... args) {
-    if (m_TraceEnabled)
+    bool enableTrace = PI_TRACE_ENABLED;
```

Comment:
It is still a getenv() per each PI call, which is redundant. Please make this a static member of the class and initialize just once.